### PR TITLE
Migrate pt-run to a complete Lua frontend

### LINKED ISSUE #13: feature request: turn pt-run into pt-lua?
Maybe we could copy-paste the [lua.c](https://github.com/lua/lua/blob/master/lua.c) from upstream Lua, and just change the pcall at the bottom, to insert our traceback function? This way, pt-run would become a drop-in replacement for /usr/bin/lua.

### PR DESCRIPTION
**Pull request by singul4ri7y**
Changelog: 
 - Migrate `pt-run` to `pt-lua`, which acts like a complete Lua frontend with our custom traceback function being the default one.
 - Add `-isystem` flag in `CPPFLAGS` of Makefile, because if Lua and Lua internals are installed in the same machine, sometimes compiler finds it troublesome regarding which Include directory to use, /usr for Lua or /usr/local for Lua internals. Maybe don't need this if Lua internals always copes up with the upstream Lua version, but I would like to keep it just incase. Noteworthy to mention, using `-I` won't do.
 - Update tracebacks tests, removing the change directory to each tests. Makes the tracebacks inconsistent (module traces having path prefix where Lua traces does not) in the tests and also unnecessary now that we don't have separate Makefiles to take care of.
 
Resolves #13

### CODE REVIEW

**Comment by hugomg** [comment]
Since renaming pt-run is an API change that will break Pallene builds, can we implement pallene-tracer version numbers before we merge this?

File: Makefile
Comment (hugomg) [inline Makefile]: Why doesn't the -I work? This -isystem is very strange because it's *not* a system headerfile.

**Comment by singul4ri7y** [comment]
Let's do it after we are done stabilizing it? We are almost done tbh

**Comment by singul4ri7y** [comment]
I think we better merge this after Lua internals and Pallene is bumped up to Lua 5.4.7.

File: pt-lua.c
Comment (hugomg) [inline pt-lua.c]: No need to use luai_likely here. This macro has marginal benefit and only in the guts of the hottest of hot loops. Not the case here.

File: pt-lua.c
Comment (hugomg) [inline pt-lua.c]: Suggestion:

    if (should_print) {
        addvalue()
    } else {
        pop();
        if (ellipsis) {
            print the skipped frames message
            ellipsis = false
        }
    }

File: pt-lua.c
Comment (hugomg) [inline pt-lua.c]: This function needs a more descriptive name & comment.

Two of the arguments are mutable state, passed via pointers. Can we avoid that or at least reduce the amount of mutable state?

1. I think we could increment the pframes outside of makeitcount. Yes it's "code duplication" but I think it will be easier to understand than a pointer.
2. We can deduce ellipsis boolean similarly to what we did for should_print. If the frame number is exactly equal to the smallest integer that would not be printed, then we print the ellipsis message.
3. Give longer names to pframes and nframes, because both are "number of frames". It sounds like pframe is current frame and nframe is total frames?

File: pt-lua.c
Comment (hugomg) [inline pt-lua.c]: Just call makeitcount directly, without the macro. The macro obscures the arguments of the function.

File: pt-lua.c
Comment (hugomg) [inline pt-lua.c:318]: \n at the start of a printf are always a bit suspicious. Does luaL_traceback also do this?

File: pt-lua.c
Comment (hugomg) [inline pt-lua.c:352]: Please add a big comment here, explainig that everything below is copy pasted from lua.c (except for the xpcall that you changed). Don't forget to add the appropriate author & copyright information.

**Review by hugomg — CHANGES_REQUESTED** [review]

**Comment by hugomg** [comment]
Agreed on merging this after 5.4.7

As for the version numbers, I suppose we might as well do it now? We already had unrelated Pallene PRs breaking because of ptracer API changes.

File: Makefile
Comment (singul4ri7y) [reply Makefile]: This is so that compiler treats the Lua include dir as a system dir (increase the resolving priority). Say, for example, we are building against Lua internals, which has it's headers files in `/usr/local/include` directory. Now, if we already have Lua installed system-wide, the compiler would resolve the header files in `/usr/include` rather than `/usr/local/include`, because system headers (specifically /usr/include) are resolved first, then /usr/local/include headers, then headers specified by the `-I` flag and finally current directory (If included using quotes).

Problem here is, when you have different version of Lua installed in `/usr/local`, compiler would include the Lua headers from `/usr/include` instead of `/usr/local/include` because of the priority. The `-isystem` flag bumps up the priority and what I can tell, it bumps up the priority to the maximum.

Besides, everything in the `/usr` directory is part of **SYSTEM**, including `/usr/local` because binaries installed in `/usr` or `/usr/local/` can be used by all the users in system. `/usr/local` is just used by softwares which does not use standard package manager.

File: pt-lua.c
Comment (singul4ri7y) [reply pt-lua.c:318]: Lua does it this way. This is because the message we returned are appended with a last `\n` at the end and then printed.

**Comment by singul4ri7y** [comment]
Yes, let's do version tagging after this merge. I will come up with commits fixing the nits tomorrow. We won't be releasing on this tag though. Also it would be wise to mark the tag as alpha if possible.

**Comment by hugomg** [comment]
The problem is that since we switched Pallene to use the separate pallene tracer repo, any changes we make here break the CI on all the Pallene PRs, including that one that Gabriel is working on. Can we please do the version numbers first?

**Comment by singul4ri7y** [comment]
Fine by me. But it would have been great if we had atleast pt-lua. If version numbers can wait one more day, I can complete this PR and then can do version numbering.

File: pt-lua.c
Comment (singul4ri7y) [reply pt-lua.c]: 1. Sure.
2. Actually we can't. The deduction part is easy and obvious, but how would we know whether have we printed the ellipsis message? Introducing another state for that is waste and even if we do, we have to pass it to makeitcount regardless. By the way, `makeitcount` is now `render`, which makes more sense in this case. I have found out a cleaner way to solve the problem I guess.
3. Correct. pframes = Printed frames or number of frames. nframes is self explanatory.

Comment (hugomg) [reply pt-lua.c]: 2) Suppose that we have 100 frames, and the math tells to print frames 1-10 and 90-100, skipping over 11-89. We can print the "..." message in place of frame 11 and print nothing in place of 12-89. No need for a mutable flag.

File: Makefile
Comment (hugomg) [reply Makefile]: https://gcc.gnu.org/onlinedocs/gcc/Directory-Options.html

According to the gcc manual, -I should have higher priority than -isystem

> You can use -I to override a system header file, substituting your own version, since these directories are searched before the standard system header file directories. However, you should not use this option to add directories that contain vendor-supplied system header files; use -isystem for that.

/usr/local is not "vendor supplied"

Comment (singul4ri7y) [reply Makefile]: While making with command `make LUA_PREFIX=/usr` and GCC verbose output using `-v -e` flags.

With `-I`: 
```
cc -DPT_DEBUG -g -std=c99 -pedantic -Wall -Wextra -Wformat-security -I/usr/include -I. -v -e  -L/usr/lib -Wl,-E pt-lua.c -o pt-lua -llua -lm
Using built-in specs.
COLLECT_GCC=cc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-pc-linux-gnu/14.2.1/lto-wrapper
Target: x86_64-pc-linux-gnu
Configured with: /build/gcc/src/gcc/configure --enable-languages=ada,c,c++,d,fortran,go,lto,m2,objc,obj-c++,rust --enable-bootstrap --prefix=/usr --libdir=/usr/lib --libexecdir=/usr/lib --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=https://gitlab.archlinux.org/archlinux/packaging/packages/gcc/-/issues --with-build-config=bootstrap-lto --with-linker-hash-style=gnu --with-system-zlib --enable-__cxa_atexit --enable-cet=auto --enable-checking=release --enable-clocale=gnu --enable-default-pie --enable-default-ssp --enable-gnu-indirect-function --enable-gnu-unique-object --enable-libstdcxx-backtrace --enable-link-serialization=1 --enable-linker-build-id --enable-lto --enable-multilib --enable-plugin --enable-shared --enable-threads=posix --disable-libssp --disable-libstdcxx-pch --disable-werror
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 14.2.1 20240805 (GCC)
COLLECT_GCC_OPTIONS='-D' 'PT_DEBUG' '-g' '-std=c99' '-Wpedantic' '-Wall' '-Wextra' '-Wformat-security' '-I' '/usr/include' '-I' '.' '-v' '-e' '-L/usr/lib' '-o' 'pt-lua' '-mtune=generic' '-march=x86-64'
 /usr/lib/gcc/x86_64-pc-linux-gnu/14.2.1/cc1 -quiet -v -I /usr/include -I . -D PT_DEBUG pt-lua.c -quiet -dumpbase pt-lua.c -dumpbase-ext .c -mtune=generic -march=x86-64 -g -Wpedantic -Wall -Wextra -Wformat-security -std=c99 -version -o /tmp/ccBAFYs6.s
GNU C99 (GCC) version 14.2.1 20240805 (x86_64-pc-linux-gnu)
	compiled by GNU C version 14.2.1 20240805, GMP version 6.3.0, MPFR version 4.2.1, MPC version 1.3.1, isl version isl-0.26-GMP

GGC heuristics: --param ggc-min-expand=100 --param ggc-min-heapsize=131072
ignoring nonexistent directory "/usr/lib/gcc/x86_64-pc-linux-gnu/14.2.1/../../../../x86_64-pc-linux-gnu/include"
ignoring duplicate directory "/usr/include"
  as it is a non-system directory that duplicates a system directory
#include "..." search starts here:
#include <...> search starts here:
 .
 /usr/lib/gcc/x86_64-pc-linux-gnu/14.2.1/include
 /usr/local/include
 /usr/lib/gcc/x86_64-pc-linux-gnu/14.2.1/include-fixed
 /usr/include
End of search list.
Compiler executable checksum: faa3163d33b78b77071c76eebeab3034
pt-lua.c: In function ‘dolibrary’:
pt-lua.c:531:31: error: ‘LUA_IGMARK’ undeclared (first use in this function)
  531 |     suffix = strchr(modname, *LUA_IGMARK);  /* look for a suffix mark */
      |                               ^~~~~~~~~~
pt-lua.c:531:31: note: each undeclared identifier is reported only once for each function it appears in
make: *** [Makefile:71: pt-lua] Error 1
```

With `-isystem`: 
```
cc -DPT_DEBUG -g -std=c99 -pedantic -Wall -Wextra -Wformat-security -isystem/usr/include -I. -v -e  -L/usr/lib -Wl,-E pt-lua.c -o pt-lua -llua -lm
Using built-in specs.
COLLECT_GCC=cc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-pc-linux-gnu/14.2.1/lto-wrapper
Target: x86_64-pc-linux-gnu
Configured with: /build/gcc/src/gcc/configure --enable-languages=ada,c,c++,d,fortran,go,lto,m2,objc,obj-c++,rust --enable-bootstrap --prefix=/usr --libdir=/usr/lib --libexecdir=/usr/lib --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=https://gitlab.archlinux.org/archlinux/packaging/packages/gcc/-/issues --with-build-config=bootstrap-lto --with-linker-hash-style=gnu --with-system-zlib --enable-__cxa_atexit --enable-cet=auto --enable-checking=release --enable-clocale=gnu --enable-default-pie --enable-default-ssp --enable-gnu-indirect-function --enable-gnu-unique-object --enable-libstdcxx-backtrace --enable-link-serialization=1 --enable-linker-build-id --enable-lto --enable-multilib --enable-plugin --enable-shared --enable-threads=posix --disable-libssp --disable-libstdcxx-pch --disable-werror
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 14.2.1 20240805 (GCC)
COLLECT_GCC_OPTIONS='-D' 'PT_DEBUG' '-g' '-std=c99' '-Wpedantic' '-Wall' '-Wextra' '-Wformat-security' '-isystem' '/usr/include' '-I' '.' '-v' '-e' '-L/usr/lib' '-o' 'pt-lua' '-mtune=generic' '-march=x86-64'
 /usr/lib/gcc/x86_64-pc-linux-gnu/14.2.1/cc1 -quiet -v -I . -D PT_DEBUG -isystem /usr/include pt-lua.c -quiet -dumpbase pt-lua.c -dumpbase-ext .c -mtune=generic -march=x86-64 -g -Wpedantic -Wall -Wextra -Wformat-security -std=c99 -version -o /tmp/ccFvf3DJ.s
GNU C99 (GCC) version 14.2.1 20240805 (x86_64-pc-linux-gnu)
	compiled by GNU C version 14.2.1 20240805, GMP version 6.3.0, MPFR version 4.2.1, MPC version 1.3.1, isl version isl-0.26-GMP

GGC heuristics: --param ggc-min-expand=100 --param ggc-min-heapsize=131072
ignoring nonexistent directory "/usr/lib/gcc/x86_64-pc-linux-gnu/14.2.1/../../../../x86_64-pc-linux-gnu/include"
ignoring duplicate directory "/usr/include"
#include "..." search starts here:
#include <...> search starts here:
 .
 /usr/include
 /usr/lib/gcc/x86_64-pc-linux-gnu/14.2.1/include
 /usr/local/include
 /usr/lib/gcc/x86_64-pc-linux-gnu/14.2.1/include-fixed
End of search list.
Compiler executable checksum: faa3163d33b78b77071c76eebeab3034
COLLECT_GCC_OPTIONS='-D' 'PT_DEBUG' '-g' '-std=c99' '-Wpedantic' '-Wall' '-Wextra' '-Wformat-security' '-isystem' '/usr/include' '-I' '.' '-v' '-e' '-L/usr/lib' '-o' 'pt-lua' '-mtune=generic' '-march=x86-64'
 as -v -I . --gdwarf-5 --64 -o /tmp/ccuzTavW.o /tmp/ccFvf3DJ.s
GNU assembler version 2.43.0 (x86_64-pc-linux-gnu) using BFD version (GNU Binutils) 2.43.0
COMPILER_PATH=/usr/lib/gcc/x86_64-pc-linux-gnu/14.2.1/:/usr/lib/gcc/x86_64-pc-linux-gnu/14.2.1/:/usr/lib/gcc/x86_64-pc-linux-gnu/:/usr/lib/gcc/x86_64-pc-linux-gnu/14.2.1/:/usr/lib/gcc/x86_64-pc-linux-gnu/
LIBRARY_PATH=/usr/lib/gcc/x86_64-pc-linux-gnu/14.2.1/:/usr/lib/gcc/x86_64-pc-linux-gnu/14.2.1/../../../../lib/:/lib/../lib/:/usr/lib/../lib/:/usr/lib/gcc/x86_64-pc-linux-gnu/14.2.1/../../../:/lib/:/usr/lib/
COLLECT_GCC_OPTIONS='-D' 'PT_DEBUG' '-g' '-std=c99' '-Wpedantic' '-Wall' '-Wextra' '-Wformat-security' '-isystem' '/usr/include' '-I' '.' '-v' '-e' '-L/usr/lib' '-o' 'pt-lua' '-mtune=generic' '-march=x86-64' '-dumpdir' 'pt-lua.'
 /usr/lib/gcc/x86_64-pc-linux-gnu/14.2.1/collect2 -plugin /usr/lib/gcc/x86_64-pc-linux-gnu/14.2.1/liblto_plugin.so -plugin-opt=/usr/lib/gcc/x86_64-pc-linux-gnu/14.2.1/lto-wrapper -plugin-opt=-fresolution=/tmp/cca7eQxR.res -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lgcc_s -plugin-opt=-pass-through=-lc -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lgcc_s --build-id --eh-frame-hdr --hash-style=gnu -m elf_x86_64 -dynamic-linker /lib64/ld-linux-x86-64.so.2 -pie -o pt-lua -e -L/usr/lib /usr/lib/gcc/x86_64-pc-linux-gnu/14.2.1/../../../../lib/Scrt1.o /usr/lib/gcc/x86_64-pc-linux-gnu/14.2.1/../../../../lib/crti.o /usr/lib/gcc/x86_64-pc-linux-gnu/14.2.1/crtbeginS.o -L/usr/lib/gcc/x86_64-pc-linux-gnu/14.2.1 -L/usr/lib/gcc/x86_64-pc-linux-gnu/14.2.1/../../../../lib -L/lib/../lib -L/usr/lib/../lib -L/usr/lib/gcc/x86_64-pc-linux-gnu/14.2.1/../../.. -E /tmp/ccuzTavW.o -llua -lm -lgcc --push-state --as-needed -lgcc_s --pop-state -lc -lgcc --push-state --as-needed -lgcc_s --pop-state /usr/lib/gcc/x86_64-pc-linux-gnu/14.2.1/crtendS.o /usr/lib/gcc/x86_64-pc-linux-gnu/14.2.1/../../../../lib/crtn.o
/sbin/ld: warning: cannot find entry symbol -L/usr/lib; defaulting to 0000000000002440
COLLECT_GCC_OPTIONS='-D' 'PT_DEBUG' '-g' '-std=c99' '-Wpedantic' '-Wall' '-Wextra' '-Wformat-security' '-isystem' '/usr/include' '-I' '.' '-v' '-e' '-L/usr/lib' '-o' 'pt-lua' '-mtune=generic' '-march=x86-64' '-dumpdir' 'pt-lua.'
```

Please look at the header search directories. In my machine I have Lua internals version 5.4.6 installed in /usr/local and Lua 5.4.7 installed system-wide.

Either I am crazy, or the GCC version of mine or the GCC docs. Probably the latter.

Comment (hugomg) [reply Makefile]: Why are you compiling with PREFIX=/usr?

Should be PREFIX=/usr/local

Comment (singul4ri7y) [reply Makefile]: Not the point actually. Point is our desired headers are only properly resolved when we use `-isystem` instead of `-I`.

File: pt-lua.c
Comment (singul4ri7y) [reply pt-lua.c]: Got your point, fixed.

File: pt-lua.c
Comment (hugomg) [inline pt-lua.c]: I think a static variable would not work the second time, if we were to print more than one traceback?

File: pt-lua.c
Comment (hugomg) [inline pt-lua.c]: As we discussed previously, we should not need a mutable variable for this in the first place. I think we just need to check if `pframes == TOP_THRESHOLD + 1`?

**Review by hugomg — CHANGES_REQUESTED** [review]
I'm still not convinced about the -isystem. I have never seen that in a makefile. The -I ought to have been sufficient? What do you mean by headers are not properly resolved? Have you tested using LUA_PREFIX=/usr/local?

**Comment by singul4ri7y** [comment]
Have I tested with `LUA_PREFIX=/usr/local`? Yes.

It **does not** work, simple as that. There is nothing to be convinced about. the `-I` prefix supposed to append the dir to the leftmost (or topmost) header searching path string, but clearly the verbose logging I showed you earlier does not follow that. I also have double checked for multiple times, it does not work.

File: pt-lua.c
Comment (singul4ri7y) [reply pt-lua.c]: We won't be printing more than one tracebacks, because this time the interpreter going to exit :). Read the comments please.

File: pt-lua.c
Comment (singul4ri7y) [reply pt-lua.c]: We need a mutable state ensuring we have printed the "skipped frame" message. The ellipsis variable just do that for us, also telling us should we print ellipsis and if we should is it already printed. Two birds one stone.

**Comment by hugomg** [comment]
In the verbose logging, you use `-I/usr/include` and `-L/usr/lib` compilation flags. Why wasn't it /usr/local?

File: pt-lua.c
Comment (hugomg) [reply pt-lua.c]: Asif, the only mutable state we need is the pframes variable.

File: pt-lua.c
Comment (hugomg) [reply pt-lua.c]: Wait, but that isn't what Lua does? The regular stack trace can be caught by pcall and we keep going after that.

**Comment by singul4ri7y** [comment]
@hugomg Again, that's not the point. I used it because in my case if I wanted to compile pt-lua against my system Lua. It was still grabbing Lua internal headers from /usr/local/include, the -I was not working. So, I had to switch to -isystem for that. Now, the situation can be very well vice versa, say you have a backdated Lua in your system that you want to compile against and a newer version of Lua is installed in /usr/local.

By the way, using /usr/local was still giving me errors, because as usual with -isystem the /usr/local was getting resolved first.

**Comment by singul4ri7y** [comment]
I highly suggest that you try the -isystem flag yourself. Also try having Lua internals 5.4.6 with you, that would make my point aloud.

**Comment by hugomg** [comment]
That is not a configuration that we have to support. The C header file include system is not designed to handle using ptracer from usr/local but then not using the Lua from /usr/local. If the user really wants to do that, they should install lua-internals and ptracer to different prefixes.

**Comment by singul4ri7y** [comment]
I am not convinced. The `-I` flag is not clearly working as the documentation suggested. Moreover, users won't be using our Makefile to build other projects, so I think the flag is fine. If the -I flag works as intended, we may change it back.

As for prefixes. I am pretty much sure users won't like to make their life a hassle by installing to a different prefix and adding it to their path just for a single goddang flag, which is not total taboo.

File: pt-lua.c
Comment (singul4ri7y) [reply pt-lua.c]: Okay, then how would we know if we have printed the skip frames message? We can't keep printing the message.

File: pt-lua.c
Comment (singul4ri7y) [reply pt-lua.c]: Actually we are at the top-order pcall.

In Lua or PT Lua I should say,
First pcall -> calling `pmain` dispatching with all the arguments. No traceback replacement (we can't/won't in this case).
Second pcall -> in dofile() function, which gets called when the main chunk has been parsed by Lua compiler. This is where we replace the traceback. So there is no way anyone can catch the traceback with pcall at this point. Even if the Lua code manages to pcall, it won't be with our traceback, because we are not supplying it anymore.

File: pt-lua.c
Comment (singul4ri7y) [reply pt-lua.c]: Ah, got it.

Comment (hugomg) [reply pt-lua.c]: I already gave the suggestion in several previous comments. Print the ellipsis message when pframe==threshold+1 and print nothing otherwise.

We don't need a boolean flag, because we already have an incrementing counter.

**Review by hugomg — APPROVED** [review]
Consider adding version numbers before merging this.

**Comment by singul4ri7y** [comment]
Should we create version tag after merging this? Because regardless, we have to make changes to Pallene with another PR.

**Comment by hugomg** [comment]
We're overdue for a version tag Em 14 de ago. de 2024 9:24 AM, SD Asif Hossein ***@***.***> escreveu:
Should we create version tag after merging this? Because regardless, we have to make changes to Pallene with another PR.

**Pull request merged**
